data.ts

```diff
@@ -44,6 +45,10 @@ export async function populate(path: string, opts: ClientOptions) {
   const data = await getData(path)
   if (data) {
     const client = new Client(opts)
-    await client.bulk({ body: data })
+    const batch_size = 10
+    const batches = chunk(data, batch_size)
+    batches.forEach(async (batch) => {
+      await client.bulk({ body: batch })
+    })
```

Comment:
This will make all of the requests _simultaneously_ rather than sequentially. Do you see why?

Try a `for...of` loop instead.

Reply:
Do you understand why the old version executed the requests simultaneously?

Reply:
Yes, I believe the forEach method doesn't wait for the promise returned by client.bulk() to resolve before starting the next iteration.

Reply:
Do you understand why the `for...of` version does execute the requests sequentially?

Why would it be a problem to send the requests simultaneously?

Reply:
From what I understand the for..of loop waits for each await operation to finish running, letting the client.bulk() method resolve before going to the next iteration. Meanwhile forEach() just runs all asynchronous operations together without waiting for the await to finish running.

So I guess for..of is more like a loop while forEach isn't.

I'm not really sure why it'd be a problem to send the requests simultaneously. If anything I thought it would run faster. 

Do you think we should try forEach again?

Reply:
> From what I understand the for..of loop waits for each await operation to finish running, letting the client.bulk() method resolve before going to the next iteration. Meanwhile forEach() just runs all asynchronous operations together without waiting for the await to finish running.

Here's the original `forEach` version:

```ts
batches.forEach(async (batch) => {
  await client.bulk({ body: batch })
})
```

You're passing an anonymous function or closure to `batches.forEach`. Let's put that function on a separate line for a moment:

```ts
async function processBatch(batch) {
  await client.bulk({ body: batch })
}

batches.forEach(processBatch)
```

If we convert this `forEach` call to a `for...of` loop, it looks like this:

```ts
async function processBatch(batch) {
  await client.bulk({ body: batch })
}

for (const batch of batches) {
  processBatch(batches)
}
```

Now, it's clear that nothing awaits on the promise returned by `processBatch`, so the loop goes on to the next iteration without waiting for the promise to complete. In fact the entire function may return before _any_ of the API calls have completed.

> I'm not really sure why it'd be a problem to send the requests simultaneously. If anything I thought it would run faster. Maybe it overloads the server with too many requests?

Right. Presumably the original bulk call was failing because the server has rate limits.

Reply:
That makes sense. Thanks.